shadowsocks/udp.go

```diff
@@ -119,7 +119,9 @@ func (s *udpService) Start() {
 					logger.Debugf("UDP Error: %v: %v", connError.Message, connError.Cause)
 					status = connError.Status
 				}
-				s.m.AddUDPPacketFromClient(clientLocation, keyID, status, clientProxyBytes, proxyTargetBytes, timeToCipher)
+				if s.isRunning {
```

Comment:
Why does this need to be under `s.isRunning`?

Reply:
I discovered this issue after writing the test.  Shutting down the proxy causes a spurious call to `s.m.AddUDPPacketFromClient("", "", "OK", 0, 0, 0)`, after `s.clientConn.ReadFrom(...)` returns an error.

Reply:
I see. It's hard to tell that's needed. Maybe we need better error management here.

Instead of returning nil when the loop is interrupted, maybe we can return some sort of interrupted/stopped error instead, and call `s.m.AddUDPPacketFromClient` and the debug only if it's not the interrupted error.

We can also consider playing with the defer and function calls structure.

Reply:
I reordered the deferred reporting to solve this problem.